js/modules/k6/http/error_codes.go

```diff
@@ -0,0 +1,137 @@
+package http
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"runtime"
+	"syscall"
+
+	"github.com/loadimpact/k6/lib/netext"
+	"github.com/pkg/errors"
+	"golang.org/x/net/http2"
+)
+
+type errCode uint32
+
+const (
+	// non specific
+	defaultErrorCode          errCode = 1000
+	defaultNetNonTCPErrorCode errCode = 1010
+	// DNS errors
+	defaultDNSErrorCode    errCode = 1100
+	dnsNoSuchHostErrorCode errCode = 1101
+	blackListedIPErrorCode errCode = 1110
+	// tcp errors
+	defaultTCPErrorCode      errCode = 1200
+	tcpBrokenPipeErrorCode   errCode = 1201
+	netUnknownErrnoErrorCode errCode = 1202
+	tcpDialErrorCode         errCode = 1210
+	tcpDialTimeoutErrorCode  errCode = 1211
+	tcpDialRefusedErrorCode  errCode = 1212
+	tcpResetByPeerErrorCode  errCode = 1220
+	// TLS errors
+	defaultTLSErrorCode           errCode = 1300
+	x509UnknownAuthorityErrorCode errCode = 1310
+	x509HostnameErrorCode         errCode = 1311
+
+	// HTTP2 errors
+	// defaultHTTP2ErrorCode errCode = 1600 // commented because of golint
+	// HTTP2 GoAway errors
+	unknownHTTP2GoAwayErrorCode errCode = 1610
+	// errors till 1611 + 13 are other HTTP2 GoAway errors with a specific errCode
+
+	// HTTP2 Stream errors
+	unknownHTTP2StreamErrorCode errCode = 1630
+	// errors till 1631 + 13 are other HTTP2 Stream errors with a specific errCode
+
+	// HTTP2 Connection errors
+	unknownHTTP2ConnectionErrorCode errCode = 1650
+	// errors till 1651 + 13 are other HTTP2 Connection errors with a specific errCode
+)
+
+var (
+	tcpResetByPeerErrorCodeMsg  = "write: connection reset by peer"
+	tcpDialTimeoutErrorCodeMsg  = "dial: i/o timeout"
+	tcpDialRefusedErrorCodeMsg  = "dial: connection refused"
+	tcpBrokenPipeErrorCodeMsg   = "write: broken pipe"
+	netUnknownErrnoErrorCodeMsg = "%s: unknown errno `%d` on %s with message `%s`"
+	dnsNoSuchHostErrorCodeMsg   = "lookup: no such host"
+	blackListedIPErrorCodeMsg   = "ip is blacklisted"
+	http2GoAwayErrorCodeMsg     = "http2: received GoAway with http2 ErrCode %s"
+	http2StreamErrorCodeMsg     = "http2: stream error with http2 ErrCode %s"
+	http2ConnectionErrorCodeMsg = "http2: connection error with http2 ErrCode %s"
+	x509HostnameErrorCodeMsg    = "x509: certificate doesn't match hostname"
+)
+
+func http2ErrCodeOffset(code http2.ErrCode) errCode {
+	if code > http2.ErrCodeHTTP11Required {
+		return 0
+	}
+	return 1 + errCode(code)
+}
+
+// returns the errorCode and a specific error message for given error. If errror message is empty
+// than the original error string should be used
+func errorCodeForError(err error) (errCode, string) {
```

Comment:
I agree with all of those except that we are not there yet and at this point I think it is better if we don't optimise for what is needed in the future. It is specifically this that I moved it to a separate file and is totally separate from everything, but as is currently not required by anything I will be making it a separate module with the idea that one day someone will remember to use it ...

Reply:
Yeah, but we know that we'd like to soon reuse this code for other existing functionality (websocket errors, when we fix their metrics emission - last part of https://github.com/loadimpact/k6/issues/885#issuecomment-450642259), and things like gRPC and other protocols, which we'd like to implement as soon as possible. So it's not refactoring for some nebulous future, it'd be needed for stuff that we plan to soon implement. Even worse, other users may want to implement them, and having this here will be in the way...

Since if we or any users want to implement one of those things, and this code is still in the `js/http` package, it would likely first have to moved in a new package in one PR, with the actual new feature in another PR, which seems like a lot of wasted effort and an obstacle to contributors. Especially if we do any changes here in the meantime, that depend on the local package nature, and which would have to be fixed before refactoring the package in the future.

Reply:
Also, I think the philosophical point is somewhat moot, since it seems like we may actually need to do this now anyway... I decided to look into how we can add the `error_code` metric tag, since I didn't see how we add the `error` tag to HTTP metrics currently. I didn't find it in this package, it happens here instead: https://github.com/loadimpact/k6/blob/master/lib/netext/transport.go#L57-L62

I think the `error` tag got moved there in the fix to properly measure HTTP requests when there are redirects. And it makes some sense, since an HTTP request can have a successful redirect (no error) to a URL that returns an error. And we need to emit both of those sets of metrics. But that also means that the `error` metric tag won't be fixed by this PR, it will still contain highly variable strings...

Reply:
Regarding that last observation: there should also probably be a test in this PR that actually checks this at the engine level, by making one or more broken HTTP requests, capturing the metrics and seeing if the `error` and `error_code` tags look like we want them to look

Reply:
Testing it that way was my initial idea but decided that I will do it only for one or two errors that will be easier to reproduce :) like blacklist Ip and let say an unresolvable address as reproducing the majority of the errors will be kind of hard :(